pkg/ingester/ingester_v2.go

```diff
@@ -272,6 +249,78 @@ func (i *Ingester) v2LabelNames(ctx old_ctx.Context, req *client.LabelNamesReque
 	}, nil
 }
 
+func (i *Ingester) v2MetricsForLabelMatchers(ctx old_ctx.Context, req *client.MetricsForLabelMatchersRequest) (*client.MetricsForLabelMatchersResponse, error) {
+	userID, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	db := i.getTSDB(userID)
+	if db == nil {
+		return &client.MetricsForLabelMatchersResponse{}, nil
+	}
+
+	// Parse the request
+	from, to, matchersSet, err := client.FromMetricsForLabelMatchersRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new instance of the TSDB querier
+	q, err := db.Querier(int64(from), int64(to))
+	if err != nil {
+		return nil, err
+	}
+	defer q.Close()
+
+	// Run a query for each matchers set and collect all the results
+	added := map[string]struct{}{}
+	result := &client.MetricsForLabelMatchersResponse{
+		Metric: make([]*client.Metric, 0),
+	}
+
+	for _, matchers := range matchersSet {
+		convertedMatchers, err := cortex_tsdb.FromLegacyLabelMatchersToMatchers(matchers)
+		if err != nil {
+			return nil, err
+		}
+
+		seriesSet, err := q.Select(convertedMatchers...)
+		if err != nil {
+			return nil, err
+		}
+
+		for seriesSet.Next() {
+			if seriesSet.Err() != nil {
+				break
+			}
+
+			// Given the same series can be matched by multiple matchers and we want to
+			// return the unique set of matching series, we do check if the series has
+			// already been added to the result
+			ls := seriesSet.At().Labels()
+			key := ls.String()
+			if _, ok := added[key]; ok {
```

Comment:
Just to be consistent with upstream, can we use this: https://github.com/prometheus/prometheus/blob/0ae4899c47fbcdab1b0bebedf213b4424f50f760/web/api/v1/api.go#L546-L550

It's an exposed function and I see it fitting perfectly here.

Reply:
I checked it. The `storage.NewMergeSeriesSet()` you pointed out works with the `storage` pkg structs, while here we work with `tsdb` structs (incompatible types). I checked the merge functions in `tsdb` pkg and looks a bit overkilled for what we need here: they merge series samples, while here we just need a unique set of labels.

_I would suggest to keep this code as is._